### PR TITLE
Merge RunPeriod and RunPeriod:CustomRange

### PR DESCRIPTION
**Pull request by jasondegraw**
Pull request overview
---------------------
Merge the RunPeriod and RunPeriod:CustomRange into a single input object and modify the processing of input so that there is always a year associated with a run period.

### Work Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
   - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
   - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
   - NewFeature: This pull request includes code to add a new feature to EnergyPlus
   - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
   - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog

### Review Checklist
This will not be exhaustively relevant to every PR.
 - [x] Code style (parentheses padding, variable names)
 - [x] Functional code review (it has to work!)
 - [ ] If defect, results of running current develop vs this branch should exhibit the fix
 - [ ] CI status: all green or justified
 - [x] Performance: CI Linux results include performance check -- verify this
 - [x] Unit Test(s)
 - C++ checks:
   - [x] Argument types
   - [x] If any virtual classes, ensure virtual destructor included, other things
 - IDD changes:
   - [x] Verify naming conventions and styles, memos and notes and defaults
   - [x] Open windows IDF Editor with modified IDD to check for errors
   - [x] If transition, add rules to spreadsheet
   - [x] If transition, add transition source
   - [x] If transition, update idfs
 - [x] Documentation changes in place
 - [x] Changed docs build successfully

### CODE REVIEW

**Comment by jasondegraw** [comment]
@Myoldmopar @mjwitte I believe all the diffs here are expected, but since develop has moved again it's likely that merging in current develop will cause failures that need to be fixed.

I was pretty aggressive with this and eliminated RunPeriod:CustomRange completely. It wouldn't be too difficult to add it back in and mark it as deprecated (or something). Changing that will not cause diffs, as there were/are no tests of RunPeriod:CustomRange.

**Comment by nrel-bot-3** [comment]
@jasondegraw @lgentile it has been 14 days since this pull request was last updated.

**Comment by nrel-bot-2** [comment]
@jasondegraw @lgentile it has been 14 days since this pull request was last updated.

**Comment by nrel-bot-3** [comment]
@jasondegraw @lgentile it has been 14 days since this pull request was last updated.

File: idd/Energy+.idd.in
Comment (mjwitte) [inline idd/Energy+.idd.in:1302]: It's probably a little late to ask for this, but it would be nice if this field is ignored when Begin Year is entered and just uses the correct day.

**Review by mjwitte — COMMENTED** [review]
Just reviewing the IDD changes, I'm ok with this. Does this cause any changes in output?

File: idd/Energy+.idd.in
Comment (jasondegraw) [reply idd/Energy+.idd.in:1302]: @mjwitte  Right now it will compute the correct weekday when this is left blank and if not it throws a severe error if the begin year and weekday don't match. I could change that if you feel strongly about it.

**Comment by jasondegraw** [comment]
@Myoldmopar @mjwitte Reformatting locally didn't help with WeatherManager.cc, still seeing conflicts there that shouldn't be there. Have no idea about ExternalInterface.cc.

File: idd/Energy+.idd.in
Comment (mjwitte) [reply idd/Energy+.idd.in:1302]: That works, but then it shouldn't have a hard default in the IDD and the note should be updated to say that.

**Review by mjwitte — APPROVED** [review]

File: idd/Energy+.idd.in
Comment (jasondegraw) [reply idd/Energy+.idd.in:1302]: @mjwitte I guess it's really a soft default, so I agree that should be changed. Will do that once I get this merge thing worked out.

**Comment by Myoldmopar** [comment]
@jasondegraw 
- CI looks like it will be mostly happy with the last set of runs here.  There is one file with small diffs.  They do indeed look tiny.  But do you know why they are diffs at all?
- Did you address the comment about the soft default from above?
- This needs transition rules, right?

**Comment by jasondegraw** [comment]
@Myoldmopar I'm working on getting the default handling worked out a little better right now., along with a couple other things that need to be changed. Yes, it will need transition rules, which I will need help with.

**Comment by jasondegraw** [comment]
@Myoldmopar OK, that's the last thing. The leap year behavior is going to need some tuning, but wrong weekday is now corrected with a warning.

**Comment by Myoldmopar** [comment]
@jasondegraw I added some partial functionality for the transition.  I still have questions, but it at least compiles.  I'll try to put my questions into the Google Doc as comments so we can address them there and then pull the final rules into the InputRules file here.

**Comment by jasondegraw** [comment]
@Myoldmopar Yes, let's regroup. It looks like you're having to do a lot of the same things in F90 that I'm doing in C++, and that feels wrongish.

**Comment by Myoldmopar** [comment]
Current code results in the following differences:
[before.idf.txt](https://github.com/NREL/EnergyPlus/files/2295014/before.idf.txt)
[after.idf.txt](https://github.com/NREL/EnergyPlus/files/2295015/after.idf.txt)

I tried resolving the conflicts with a merge in develop, but it wasn't immediately obvious so I'll leave that to you @jasondegraw.  If the transition code looks pretty good then this could be nearly ready to go.  I know that the IDFs being produced have extra whitespace padding around the new fields, but I feel like I'm not going to spend much time investigating that if the IDF works.

**Comment by Myoldmopar** [comment]
@jasondegraw I pushed up the last changes that you thought looked right in the email a couple hours ago.  I again did not try to resolve the conflicts.  If you could do that, it would be great to see CI turn green one more time and maybe have this be the next one to drop in.

File: src/Transition/CreateNewIDFUsingRulesV9_0_0.f90
Comment (mjwitte) [inline src/Transition/CreateNewIDFUsingRulesV9_0_0.f90:747]: @Myoldmopar `RoundSigDigits(YearNumber,0)` should solve the extra spaces problem.

File: src/Transition/CreateNewIDFUsingRulesV9_0_0.f90
Comment (mjwitte) [inline src/Transition/CreateNewIDFUsingRulesV9_0_0.f90:767]: Several places have a comment about a warning, but I don't see any code to throw one. Or do you mean EnergyPlus 9.0 will throw the warning?

File: doc/input-output-reference/src/overview/group-location-climate-weather-file-access.tex
Comment (mjwitte) [inline doc/input-output-reference/src/overview/group-location-climate-weather-file-access.tex:691]: @jasondegraw A little late in the game to suggest a change here so feel free to ignore this comment. The name of this field confuses me, because it leads me to think more about the weather data values rather than the dates. Even synthesized data could be created with valid date fields that you want to match. For example, one might create an epw with 20 years of projected weather data and want to run specific years from within that epw. Would you be open to something like "Match (or Align) Weather File Years" or "Weather Files Dates Must Match Exactly" or "Strict Weather File Dates"?

**Comment by mjwitte** [comment]
Oh, and @Myoldmopar this needs to transition RunPeriod:CustomRange into RunPeriod.

File: doc/input-output-reference/src/overview/group-location-climate-weather-file-access.tex
Comment (jasondegraw) [reply doc/input-output-reference/src/overview/group-location-climate-weather-file-access.tex:691]: @mjwitte The current name came about because it triggers the behavior more like what you used to get with RunPeriod:CustomRange that was determined by a field named "ActualWeather" or something like that. Mainly that means that the first entry of start date must appear in the EPW (and that's all) and that if a leap year is covered in the run period the EPW must have a leap day. So I'm not sure about "strict" as a descriptor. How about something like "Use actual calendar dates"? I guess that could lead to other questions, maybe something similar would get across the message.

Comment (mjwitte) [reply doc/input-output-reference/src/overview/group-location-climate-weather-file-access.tex:691]: I'm not finding an old field like that, but it's not big deal here. I'm going to take a crack at writing the transition code for RunPeriod:CustomRange.

Comment (mjwitte) [reply doc/input-output-reference/src/overview/group-location-climate-weather-file-access.tex:691]: Also, the performance tests need updating.

Comment (jasondegraw) [reply doc/input-output-reference/src/overview/group-location-climate-weather-file-access.tex:691]: @mjwitte Sorry, it wasn't a field, it is/was a member variable:

https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/WeatherManager.hh#L382

**Comment by mjwitte** [comment]
@jasondegraw @Myoldmopar I think transition is ready to go. It currently always writes all of the fields for RunPeriod in the new file. Here's a revised testfile and result.
[before.idf.txt](https://github.com/NREL/EnergyPlus/files/2298814/before.idf.txt)
[before.idfnew.txt](https://github.com/NREL/EnergyPlus/files/2298813/before.idfnew.txt)

**Review by mjwitte — CHANGES_REQUESTED** [review]
@jasondegraw When running this locally, day of week always starts with 1. Run 5ZoneAirCooled, as-is with Tuesday, and with Wednesday, Jan 1 is day type 1 in both, should be 3 and 4. The day type does increment foward from there. Also ran a test with year 2019, start day blank, still starts on day type 1.

**Comment by mjwitte** [comment]
Also, the error message when Treat as Actual = Yes but the epw doesn't have years, should mention changing Treat as Actual = No might solve the problem.
Test files attached here.
[6478-RunPeriodYear.zip](https://github.com/NREL/EnergyPlus/files/2303130/6478-RunPeriodYear.zip)

**Comment by mjwitte** [comment]
Why are there [small diffs in 5ZoneAirCooledWithCoupledInGradeSlab](https://nrel.github.io/EnergyPlusBuildResults/EnergyPlus-b2ac52b5e50556d9481a3eb23321f15716b30d14-x86_64-Linux-Ubuntu-14.04-gcc-4.8.html)? Perhaps something to do with the way the GroundDomain initializes? @mitchute any thoughts?

**Comment by mitchute** [comment]
Based on the code diffs I'm seeing, this does seem odd. An init could be the issue, but that model is going to need the simulation time for the ground temp model, so it may be something related that.

**Comment by jasondegraw** [comment]
@mjwitte I just pushed a commit that fixes the start day problem and hopefully finishes up transition changes. The `USEWEATHERFILE` option is gone now, and it was still getting copied over.

**Comment by jasondegraw** [comment]
@mjwitte Oops, I missed your comment about the warning message. Which message are you referring to exactly?

**Comment by jasondegraw** [comment]
@mitchute Could it be that the start day of the week was wrong? It looks like that file was the worse case scenario in terms of being off.

**Comment by mjwitte** [comment]
@jasondegraw @mitchute Hopefully, the start day fix will resolve the ground domain diffs.  It just didn't make sense to me that this would matter when it's a design-day only simulation. 

Regarding the error message.  This is what shows if actual=yes with a tmy file (that has no years or mixed years):
```
   ** Severe  ** GetNextEnvironment: Runperiod has been entered but weatherfile DATA PERIOD does not have year included in start/end date.
   **   ~~~   ** ...to match the RunPeriod, the DATA PERIOD should be mm/dd/yyyy for both.
   ** Severe  ** GetNextEnvironment: Runperiod [mm/dd/yyyy] (Start=01/01/2013,End=12/31/2013) requested not within Data Period(s) from Weather File
   **   ~~~   ** Weather Data Period (Start=01/01/<noyear>,End=12/31/<noyear>)
   **  Fatal  ** GetNextEnvironment: Program terminates due to preceding condition.
```

**Comment by jasondegraw** [comment]
@mjwitte How about this:

```
  ** Severe  ** GetNextEnvironment: Actual weather runperiod has been entered but weatherfile DATA PERIOD does not have year included in start/end date.
   **   ~~~   ** ...to match the RunPeriod, the DATA PERIOD should be mm/dd/yyyy for both, or
   **   ~~~   ** ...set "Treat Weather as Actual" to "No".
   ** Severe  ** GetNextEnvironment: Runperiod [mm/dd/yyyy] (Start=01/01/2013,End=12/31/2013) requested not within Data Period(s) from Weather File
   **   ~~~   ** Weather Data Period (Start=01/01/<noyear>,End=12/31/<noyear>)
   **  Fatal  ** GetNextEnvironment: Program terminates due to preceding condition.
```

**Comment by mjwitte** [comment]
Getting there.  
1. Curious why some files are showing table diffs for average number of occupants. What would have changed there?
2. Tried a file from Jan 1, 2020 to Mar 31, 2020. Day types are correct until March - there's no Feb 29.
3. In the I/O Ref for "End Year' it says "Note that this could be several years after the start date (e.g. if the end date is specified as a leap day 2/28)." - should be 29.
4. The I/O Ref RunPeriod introductory paragraph should mention how leap years will be handled (or not handled).
5. Searching the I/O Ref for "leap", found this under Schedule:File. What happens if the simulation is multi-year and include a leap year?

> Note if the simulation does not have a leap year specified but the schedule file
contains 8784 hours of data, the first 8760 hours of data will be used. The schedule manager will
not know to skip the 24 hours representing February 29.

**Comment by mjwitte** [comment]
6. This object results in 5 sets of Jan 1 - Mar 31. I was expecting complete years, ending with Mar 31 in the last year. If this is the intended behavior, the I/O Ref should say so.
```
  RunPeriod,
    ,                        !- Name
    1,                       !- Begin Month
    1,                       !- Begin Day of Month
    2016,                        !- Begin Year
    3,                      !- End Month
    31,                      !- End Day of Month
    2020,                        !- End Year
    ,                 !- Day of Week for Start Day
```
7. Might be nice if the progress messages included the year "Continuing Simulation at 03/22 for RUNPERIOD 1".

**Comment by mjwitte** [comment]
8. This RunPeriod fails.
```
  RunPeriod,
    ,                        !- Name
    1,                       !- Begin Month
    1,                       !- Begin Day of Month
    2016,                        !- Begin Year
    12,                      !- End Month
    31,                      !- End Day of Month
    2020,                        !- End Year

   ** Severe  ** GetNextEnvironment: Runperiod [mm/dd] (Start=01/01,End=12/31) requested not within Data Period(s) from Weather File
   **   ~~~   ** Weather Data Period (Start=01/01/<noyear>,End=12/31/<noyear>)
   **  Fatal  ** GetNextEnvironment: Program terminates due to preceding condition.
```
[5ZoneAirCooled-NoDay-2016-2020-FullYears.idf.txt](https://github.com/NREL/EnergyPlus/files/2307045/5ZoneAirCooled-NoDay-2016-2020-FullYears.idf.txt)

Sorry to pepper with comments. Done testing for now.

**Comment by jasondegraw** [comment]
@mjwitte I think one of the recent merges has broken something. Just guessing here, but your 2, 6, and 8 are probably related to the mixture of ordinal and Julian dates. A recent PR messed with some of that, and the conflicted merge may have left something in the wrong place. 3, 4, and (maybe) 5 I'll fix here in a minute, 7 I'll take a look, then 2, 6, and 8. For 1 I have no idea, but I'll see what I can find after I figure out what's wrong elsewhere. Were you using the Chicago TMY2 for these tests?

**Comment by mjwitte** [comment]
Using Chicago TMY3? The one that's distributed in the install package.

**Comment by jasondegraw** [comment]
@mjwitte OK, that's close enough. For 2, I think that's the behavior we want.

  * For 1/1/2020 -3/31/2020 with the TMY2, I get 2/28 as Friday and then 3/1 as Saturday.
  * For 1/1/2020 -3/31/2020 with the TMY2 with an inserted 2/29, I get 2/28 as Friday, 2/29 as Saturday, and 3/1 as Sunday.

3/1 is really a Sunday, but if there's no leap year provided, pretending it's not a leap year is probably the thing to do. Were you seeing something different?

**Comment by mjwitte** [comment]
I guess if the weather file doesn't have Feb 29 data, but the year is a leap year, there's no way to do it right. Either skip 2/29 and the day types are all wrong for the remainder of the year, or repeat Feb 28 weather which isn't right either. The other option is to reject this combination and fatal out, but I suppose that should only happen if Treat as Actual = Yes. So, the current behavior is fine and probably closest to current behavior, so just add something in the I/O Ref Begin Year description to explain this.

**Comment by lklawrie** [comment]
That would depend if the  leap year indicator is set. Again most try files
are 365 days regardless of what "year" Feb is.

Linda - sent from my phone (not iPhone)

**Comment by jasondegraw** [comment]
@mjwitte Found the issues that were causing your 6 and 8, working on fixes now. That leaves 7 and 5 to go.

**Comment by mjwitte** [comment]
@jasondegraw 5 is somewhat subtle, and could be investigated as a followup issue rather than hold this up.

**Review by mjwitte — APPROVED** [review]
Current status. Resolved: 2, 3, 4, 8
Still questions:
No. 1. Change in table average occupancy
No. 5. Schedule:File - multi-year leap year behavior
No. 6. Still getting only Jan-Mar for each year
No. 7. Adding year to progress messages
No. 9. Annual output variables report as expected on 12/31 24:00 for 2016-2020 with full years. Should there be a column of years in the csv?
No. 10. RunPeriod in eio, end year is one too high.
```
Environment,RUNPERIOD 1,WeatherFileRunPeriod,01/01/2017,12/31/2018,Sunday,365,Use RunPeriod Specified Day,No,No,No,Yes,Yes
```

These can be done as a followup issue(s). Will wait for CI to finish and double-check a few things before merging.

**Comment by mjwitte** [comment]
<ack> MD renumbered my list. See [edited comment](https://github.com/NREL/EnergyPlus/pull/6478#pullrequestreview-148563906).

**Comment by jasondegraw** [comment]
@mjwitte Just quickly before our internet drops again, I've also got 7 and 10 addressed, and 6 is partly there. No. 9 is an ESO timestamp change. Every time I even think about timestamps somebody quotes me out of context and beats me up over it, so I'd prefer to leave that alone for the time being. No 1 and 5 are still open.

**Comment by jasondegraw** [comment]
@mjwitte I have identified the cause of No. 6, and should have a fix ready sometime tomorrow morning. That leaves No 1 as the remaining known issue with execution. I can look into both that and No 5 tomorrow.

File: src/EnergyPlus/WeatherManager.cc
Comment (mjwitte) [inline src/EnergyPlus/WeatherManager.cc]: These changes are introducing tiny diffs. Let's not introduce that as part of this PR. This can be explored as a separate PR after I/O freeze.

Comment (jasondegraw) [reply src/EnergyPlus/WeatherManager.cc]: Done.

**Comment by jasondegraw** [comment]
@mjwitte @Myoldmopar I've fixed the wrap-around issue (Mike's no. 6), now there are some conflicts to fix. Once that's done, should be good to go.

File: src/EnergyPlus/GroundTemperatureModeling/FiniteDifferenceGroundTemperatureModel.cc
Comment (mjwitte) [inline src/EnergyPlus/GroundTemperatureModeling/FiniteDifferenceGroundTemperatureModel.cc:264]: @jasondegraw This was failing here with an array bounds exceeded, because `NumOfDayInEnvrn` was 366 for test file PipingSystem_Underground_TwoPipe_FD_GroundTemps. That's still the case here, but this will run now. I will include this in the followup issue.
@mitchute This section is suspect, because it's looping on `NumOfDayInEnvrn`, but the average temperature calc [here](https://github.com/NREL/EnergyPlus/blob/c4de98c52a99dccf8ef85e34b9a50a5fbda0f0aa/src/EnergyPlus/GroundTemperatureModeling/FiniteDifferenceGroundTemperatureModel.cc#L350) is dividing by `NumDaysInYear`. Seems like it should be dividing by `DayOfSim`

**Comment by mjwitte** [comment]
@jasondegraw @Myoldmopar Remaining issues noted in #6926 If CI comes back clean, I think this can drop in (basic 1-yr simulations work fine) and we can clean up the details after I/O freeze, unless you see something in the list that's a must-fix now.

File: src/EnergyPlus/GroundTemperatureModeling/FiniteDifferenceGroundTemperatureModel.cc
Comment (mitchute) [reply src/EnergyPlus/GroundTemperatureModeling/FiniteDifferenceGroundTemperatureModel.cc:264]: ```DayOfSim``` should work. It doesn't look to me like this is being reset after the loop, so that should still be valid.

**Comment by mjwitte** [comment]
Mac results look good here. Will wait for linux to finish just to be sure, then merge. Windows appears to be way far behind. As soon as this is merged, I'll make a PR to remove _RefBldgSmallHotelNew2004_Chicago_annual from the regression/integration tests. That should help move things along better.

**Comment by mjwitte** [comment]
Not waiting any longer for integration coverage to finish, merging.

**Pull request merged**
